src/main/java/seedu/address/model/ContactTagIntegrationManager.java

```diff
@@ -135,31 +132,61 @@ private void removeContactsUnderTagFromAddressBook(Tag tag) {
      */
     public void deleteTagAndDirectContactsRecursive(Tag tag) {
         removeContactsUnderTagFromAddressBook(tag);
+        Set<Tag> superTagSet = tagTree.getSuperTagsOf(tag);
         tagTree.getSubTagsOf(tag).stream().forEach(subTag -> deleteTagAndDirectContactsRecursive(subTag));
         tagTree.deleteTagAndAllSubTags(tag);
+        superTagSet.stream().filter(superTag -> !hasTag(superTag))
+                .forEach(superTag -> deleteTag(superTag));
     }
 
     /**
-     * Returns a new {@code person} object which is nearly identical to {@code personToCopy} except with
-     * {@code tagRemoved} removed.
+     * Deletes the {@code person} from the addressbook.
+     * If this {@code person} is the only contact belonging to a specific tag, that tag is also deleted.
      */
-    private Person copyPersonWithoutTag(Person personToCopy, Tag tagRemoved) {
-        HashSet<Tag> newTagSet = new HashSet<>(personToCopy.getTags());
-        newTagSet.remove(tagRemoved);
-        return new Person(personToCopy.getName(),
-                personToCopy.getPhone(),
-                personToCopy.getEmail(),
-                personToCopy.getAddress(),
-                newTagSet);
+    public void deletePerson(Person person) {
+        Set<Tag> personTags = person.getTags();
+        addressBook.removePerson(person);
+        for (Tag tag : personTags) {
+            if (!hasTag(tag)) {
+                deleteTag(tag);
+            }
+        }
+    }
+
+    /**
+     * Changes the {@code person} in the addressbook with {@code editedPerson}.
+     * If there {@code person} was the only contact in a tag and {@code editedPerson} has them removed,
```

Comment:
if the*? also should this be in present tense since the first sentence is also in present tense?